src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs

```diff
@@ -279,16 +289,7 @@ public ScopedCallSite(IService key, IServiceCallSite serviceCallSite)
 
             public virtual object Invoke(ServiceProvider provider)
             {
-                object resolved;
-                lock (provider._resolvedServices)
-                {
-                    if (!provider._resolvedServices.TryGetValue(_key, out resolved))
-                    {
-                        resolved = _serviceCallSite.Invoke(provider);
-                        provider._resolvedServices.Add(_key, resolved);
-                    }
-                }
-                return resolved;
+                return provider.GetOrAdd(_key, () => _serviceCallSite.Invoke(provider));
```

Comment:
Does this create a closure that didn't exist before? Does this fix a bug or measurably improve perf?